### PR TITLE
test: add MockBtcClient

### PR DESCRIPTION
**Pull request by bap2pecs**
## Summary
https://github.com/babylonchain/babylon-da-sdk/issues/43#issuecomment-2209846568

we want to avoid using BTC mainnet RPCs in tests

this PR sets up the initial structure using DI to mock the BTCClient

## Test Plan

```
make lint
```

### CODE REVIEW

**Review by gitferry — COMMENTED** [review]
As we are introducing `BtcClient`, why not pass this interface as the parameter of the constructor of `BabylonQueryClient`? i.e.,
```
func NewClient(config *Config, btcClient BtcClient) (*BabylonQueryClient, error) {
```

This way, we can pass either the real btc client struct or the mocked one according to our need. And, we don't need to implement the mock by ourselves. Instead, we can use the third-party mock lib to generate mocks. Some example: https://github.com/babylonchain/staking-indexer/blob/ce71c0a6f792b12abf2c225305fb671dc673b976/btcscanner/btc_scanner.go#L59

**Comment by bap2pecs** [comment]
> As we are introducing BtcClient, why not pass this interface as the parameter of the constructor of BabylonQueryClient?

we can do that but why is that better? so context is we need to call `QueryIsBlockBabylonFinalized` in op-node https://github.com/babylonchain/optimism/blob/feat/babylon-rfc/op-node/rollup/finality/finalizer.go#L219-L238 as well as in our e2e test

in our test, we can pass in a mock object. but in OP's codebase, it's a bit hard since we only want to have minimal changes to their codebase.

> And, we don't need to implement the mock by ourselves. Instead, we can use the third-party mock lib to generate mocks. 

@gitferry do you mean sth like github.com/golang/mock/mockgen?

> Some example: https://github.com/babylonchain/staking-indexer/blob/ce71c0a6f792b12abf2c225305fb671dc673b976/btcscanner/btc_scanner.go#L59

what part uses mock in this example?

**Comment by gitferry** [comment]
So, in the codebase, we should always use the real implementation of the BTC client but the network parameter can be configured, e.g., `signet` or `mainnet`. We usually use `signet` for e2e test.

We use mocked btc client only for unit tests. See https://github.com/babylonchain/staking-indexer/blob/ce71c0a6f792b12abf2c225305fb671dc673b976/btcscanner/btc_scanner_test.go#L48. The mocked code is generated here https://github.com/babylonchain/staking-indexer/blob/623ddf5d7c617502c3a2dcfd79f93eef0d4046b0/Makefile#L67

**Comment by bap2pecs** [comment]
> So, in the codebase, we should always use the real implementation of the BTC client but the network parameter can be configured

the problem is our FP e2e tests use real implementation of the op-node, which uses the BTC clients here.

but we can't mock anything in the op-node

>  We usually use signet for e2e test.

this doesn't work in our FP e2e test b/c the inserted delegation is default to start from block 0. so if we use signet, we cannot control when the delegation unbounds.

**Comment by bap2pecs** [comment]
@gitferry it also feels wrong to me to query signet RPC in CI just for running e2e tests. also, we will meet with the same rate limiting issue which can fail the CI

**Comment by gitferry** [comment]
> the problem is our FP e2e tests use real implementation of the op-node, which uses the BTC clients here.
but we can't mock anything in the op-node

OK, I see your point. It's limited by the op-node. Why not also make the sdk client an interface and part of the `Finalizer`? Maybe I was still missing some context. Would be happy to jump in a call for this when you are OK

> it also feels wrong to me to query signet RPC in CI just for running e2e tests. also, we will meet with the same rate limiting issue which can fail the CI

Can we run the signet locally for e2e like this? https://github.com/babylonchain/staking-indexer/blob/d085cbfa868f7b45be9b83e72c18911087fd5773/itest/bitcoind_node_setup.go#L47

File: sdk/client.go [hunk 1]
```diff
@@ -49,7 +54,7 @@ func (config *Config) getRpcAddr() (string, error) {
 type BabylonQueryClient struct {
 	config    *Config
 	bbnClient *bbnclient.Client
-	btcClient *btcclient.BTCClient
+	BtcClient BtcClient
```
Comment (SebastianElvis) [inline sdk/client.go:57]: Why should `BtcClient` be a part of `BabylonQueryClient` rather than being standalone?

File: sdk/client.go [hunk 2]
```diff
@@ -77,15 +82,21 @@ func NewClient(config *Config) (*BabylonQueryClient, error) {
 		return nil, fmt.Errorf("failed to create Babylon client: %w", err)
 	}
 
+	var btcClient BtcClient
 	// Create BTC client
-	btcClient, err := btcclient.NewBTCClient(config.BTCConfig, logger)
+	switch config.ChainType {
+	case BabylonLocalnet:
+		btcClient, err = testutils.NewMockBTCClient(config.BTCConfig, logger)
```
Comment (SebastianElvis) [inline sdk/client.go:89]: This seems only for test purposes, so I'm not sure whether it should be in this place. This is another argumnt for moving the BTC client out of the `BabylonQueryClient`

File: testutils/mock_btc_client.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package testutils
+
+import (
+	"github.com/babylonchain/babylon-da-sdk/btcclient"
+	"github.com/stretchr/testify/mock"
+	"go.uber.org/zap"
+)
+
+// MockBtcClient is the mock implementation of BtcClient.
+type MockBtcClient struct {
```
Comment (SebastianElvis) [inline testutils/mock_btc_client.go:10]: Suggest we formalise the mock procedure for unit tests, like what we did in finality provider

File: sdk/client.go (hunk 2 above)
Comment (bap2pecs) [reply sdk/client.go:89]: just had a quick call w @gitferry. this is more like a temporary workaround to unblock the integration. after we use local signet, this would be a concern anymore. but probably need some major refactoring

File: sdk/client.go (hunk 1 above)
Comment (bap2pecs) [reply sdk/client.go:57]: the query client is the umbrella client to use the SDK. the user shouldn't need to be aware of the btc client

File: testutils/mock_btc_client.go (hunk 3 above)
Comment (bap2pecs) [reply testutils/mock_btc_client.go:10]: for sure. but I'd prefer we do that a bit later and along with https://github.com/babylonchain/babylon-da-sdk/issues/43

for now I just want to glue everything together and have this PR to unblock the rate limiting issue in CI

File: sdk/client.go (hunk 1 above)
Comment (SebastianElvis) [reply sdk/client.go:57]: Sure, although then the `BabylonQueryClient` could be probably renamed to something else as it's not just a Babylon query client

File: sdk/client.go (hunk 2 above)
Comment (SebastianElvis) [reply sdk/client.go:89]: Then how about we provide `btcClient` as an input to the constructor function of the client? This way it's up to the caller to construct whatever BTC client it wants

Comment (bap2pecs) [reply sdk/client.go:89]: yeah I also discussed that with @gitferry 

the tricky part is we can't pass a mock client from the OP codebase where it's called if we want to make the op changes minimal

File: sdk/client.go (hunk 1 above)
Comment (bap2pecs) [reply sdk/client.go:57]: fixed in upstream

File: sdk/client.go (hunk 2 above)
Comment (SebastianElvis) [reply sdk/client.go:89]: You can specify an interface for the client, let the mock client and real client implement the interface, and let the constructor function to take an interface as input, right?

Comment (bap2pecs) [reply sdk/client.go:89]: > let the constructor function to take an interface as input

then how do we mock the `btcClient` in the OP's codebase?

Comment (SebastianElvis) [reply sdk/client.go:89]: That means OP is responsible for defining the interface, not Babylon DA SDK. The interface shall be defined by the user of a concrete implementation, because the interface def represents the "view" on the implementation from the user. Wdyt?

Comment (bap2pecs) [reply sdk/client.go:89]: > That means OP is also responsible for defining the interface

it's easy to define the interface in OP codebase. but that means we will need to instantiate the client object in the OP codebase. so we need some logic there to determine whether to use mock or real client.

that increases complexities in OP's codebase which I want to avoid. that's why I add the switch statement in this repo instead of putting it in OP

Comment (SebastianElvis) [reply sdk/client.go:89]: The complexity can always be solved by defining constructor functions for the BTC client in the Babylon DA SDK. It also makes sense to me for specifying mock/real client in OP code base as this can be helpful for unit tests in OP (consider we want to test some edge cases in OP node without e2e).

But this is not a high priority and I don't have strong opinion on this, your call

**Comment by SebastianElvis** [comment]
Discussed offline and it seems that if we use interface in the OP side, we will need to change a lot of code there, which is not desired. So we decided to follow the `switch` approach in this PR

**Review by SebastianElvis — APPROVED** [review]
Lgtm!

**Comment by bap2pecs** [comment]
> Discussed offline and it seems that if we use interface in the OP side, we will need to change a lot of code there, which is not desired. So we decided to follow the `switch` approach in this PR

yeah if later we want to get this accepted by the OP core team, I imagine there will be massive refactoring so probably no need to add too much change for now
